File: 40-combination-sum-ii/40-combination-sum-ii.java
class Solution {
       private void possibleComb(int candidates[],int target,List<List<Integer>> results,int index,ArrayList<Integer> comb,int sum){
        
        if(sum == target){
            results.add(new ArrayList<Integer>(comb));
            return;
        }
         
        if(index >= candidates.length) return;
        
        for(int i=index;i<candidates.length;i++){
            if(i > index && candidates[i] == candidates[i-1]) continue;   
            
            if(sum + candidates[i] <= target){
                comb.add(candidates[i]);
                possibleComb(candidates,target,results,i + 1,comb,sum + candidates[i]);   
                comb.remove(comb.size() - 1);
            }
        }
        
    }
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
         List<List<Integer>> results = new ArrayList<>();
         Arrays.sort(candidates);
         possibleComb(candidates,target,results,0,new ArrayList<Integer>(),0);
        return results;
    }
}

File: 40-combination-sum-ii/40-combination-sum-ii.cpp
class Solution {
public:
    void helper(vector<int> &nums,int target,int ind,vector<vector<int>> &ans,vector<int> &curr){
        if(target == 0){
            ans.push_back(curr);
            return;
        }
        if(ind == nums.size() || target < 0) return;
        
        for(int j=ind;j<nums.size();j++){
            if(j > ind && nums[j] == nums[j-1]) continue;
            curr.push_back(nums[j]);
            helper(nums,target - nums[j],j+1,ans,curr);
            curr.pop_back();
        }
    }
    vector<vector<int>> combinationSum2(vector<int>& nums, int target) {
        sort(nums.begin(),nums.end());
        vector<vector<int>> ans;
        vector<int> curr;
        helper(nums,target,0,ans,curr);
        return ans;
    }
};